akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala

```diff
@@ -123,8 +125,8 @@ import scala.concurrent.{ Future, Promise }
           } else if (isAvailable(out)) {
             push(out, elem)
             promise.success(QueueOfferResult.Enqueued)
-          } else if (pendingOffer.isEmpty)
-            pendingOffer = Some(offer)
+          } else if (!pendingOffers.isFull)
+            pendingOffers.enqueue(offer)
```

Comment:
Unrelated to this PR, but why do we not complete the promise when an offer is successfully enqueued (`offer.promise.success(QueueOfferResult.Enqueued)`)?  If the queue is not full shouldn't the offer promise be immediately resolved?

Reply:
Em... I think you mixed up buffer and pending offers. I'll try to explain a bit:
-  buffer:
    - holds elements of the queue that are ready for downstream demands
    - instantly competes offer promise on enqueue
    - can be zero then pending offers will be completed only on downstream demands
- pending offers:
    - holds offers to the queue (buffer), when queue is not ready to handle element (only when buffer is full and overflow strategy is backpressure)
    - doesn't complete offer promise on enqueue to support backpressure mechanism
    - should be grater then zero to be able to work with minimum one elements producer

Reply:
Yes, I confused what was being enqueued when I made this comment.  Thanks for the clarification.